src/streamlink/plugins/nrk.py

```diff
@@ -1,64 +1,95 @@
+import logging
 import re
 from urllib.parse import urljoin
 
 from streamlink.plugin import Plugin
 from streamlink.plugin.api import validate
-from streamlink.stream import HLSStream
-
-COOKIE_PARAMS = (
-    "devicetype=desktop&"
-    "preferred-player-odm=hlslink&"
-    "preferred-player-live=hlslink"
-)
-
-_id_re = re.compile(r"/(?:program|direkte|serie/[^/]+)/([^/]+)")
-_url_re = re.compile(r"https?://(tv|radio).nrk.no/")
-_api_baseurl_re = re.compile(r'''apiBaseUrl:\s*["'](?P<baseurl>[^"']+)["']''')
-
-_schema = validate.Schema(
-    validate.transform(_api_baseurl_re.search),
-    validate.any(
-        None,
-        validate.all(
-            validate.get("baseurl"),
-            validate.url(
-                scheme="http"
-            )
-        )
-    )
-)
-
-_mediaelement_schema = validate.Schema({
-    "mediaUrl": validate.url(
-        scheme="http",
-        path=validate.endswith(".m3u8")
-    )
-})
+from streamlink.stream import HLSStream, HTTPStream
+
+log = logging.getLogger(__name__)
 
 
 class NRK(Plugin):
+    _psapi_url = 'https://psapi.nrk.no'
+    # Program type to manifest type mapping
+    _program_type_map = {
+        'direkte': 'channel',
+        'serie': 'program',
+        'program': 'program',
+        'podkast': 'podcast',
+    }
+
+    _url_re = re.compile(r"https?://(?:tv|radio)\.nrk\.no/(program|direkte|serie|podkast)(?:/.+)?/([^/]+)")
```

Comment:
this needs an update, you will catch URLs without a program_id

---

url without program_id
https://tv.nrk.no/serie/kongelige-fotografer/sesong/1/episode/2/avspiller

url with program_id
https://tv.nrk.no/serie/nytt-paa-nytt/2020/MUHH43003020

---

add new url tests in https://github.com/streamlink/streamlink/blob/master/tests/plugins/test_nrk.py

Reply:
Thanks, I somehow missed this. There doesn't appear to be a consistent `program_id` format according to the docs (they list 6 different formats), so I think the best way to handle this is to try to extract the `nrk:program-id` meta property, then fall back to the last part of the URL if not found. The fallback is needed for live streams under `/direkte/` to work.